ocaml/runtime/domain.c

```diff
@@ -1784,7 +1784,8 @@ static void handover_finalisers(caml_domain_state* domain_state)
 
   if (f->todo_head != NULL || f->first.size != 0 || f->last.size != 0) {
     /* have some final structures */
-    if (caml_gc_phase != Phase_sweep_main) {
+    if (caml_gc_phase != Phase_sweep_main &&
+        caml_gc_phase != Phase_sweep_and_mark_main) {
```

Comment:
This means that `caml_add_orphaned_finalisers` can be called in `Phase_sweep_and_mark_main` so it's assertion also needs updating. I've done that in #2358.